url/url.cpp

```diff
@@ -51,4 +101,1341 @@ std::string blob_url_create(Origin const &origin) {
+            validation_error(ValidationError::InvalidURLUnit);
+        }
+
+        if (peek() == '%'
+                && (remaining().size() < 2 || !util::is_hex_digit(remaining()[0])
+                        || !util::is_hex_digit(remaining()[1]))) {
+            validation_error(ValidationError::InvalidURLUnit);
+        }
+
+        buffer_ += peek();
+    }
+}
+
+// https://url.spec.whatwg.org/#fragment-state
+void UrlParser::state_fragment() {
+    if (!is_eof()) {
+        if (!is_url_codepoint(util::utf8_to_utf32(peek(4))) && peek() != '%') {
+            validation_error(ValidationError::InvalidURLUnit);
+        }
+
+        if (peek() == '%'
+                && (remaining().size() < 2 || !util::is_hex_digit(remaining()[0])
+                        || !util::is_hex_digit(remaining()[1]))) {
+            validation_error(ValidationError::InvalidURLUnit);
+        }
+
+        url_.fragment.value() += util::percent_encode(peek(1), PercentEncodeSet::fragment);
+    }
+}
+
+// https://url.spec.whatwg.org/#concept-domain-to-ascii
+std::optional<std::string> UrlParser::domain_to_ascii(std::string_view domain, bool be_strict) {
+    icu_init();
+
+    std::string ascii_domain;
+    icu::StringByteSink<std::string> tmp{&ascii_domain};
+
+    icu::IDNAInfo inf;
+    UErrorCode err = U_ZERO_ERROR;
+
+    std::uint32_t opts = UIDNA_NONTRANSITIONAL_TO_ASCII | UIDNA_CHECK_BIDI | UIDNA_CHECK_CONTEXTJ;
+
+    if (be_strict) {
+        opts |= UIDNA_USE_STD3_RULES;
+    }
+
+    auto uts = icu::IDNA::createUTS46Instance(opts, err);
+
+    if (U_FAILURE(err)) {
+        spdlog::info("Failed to create UTS46 instance, error {}; idna data probably missing from icu",
+                static_cast<std::int64_t>(err));
+
+        return std::nullopt;
+    }
+
+    err = U_ZERO_ERROR;
+
+    uts->nameToASCII_UTF8(domain, tmp, inf, err);
+
+    delete uts;
+
+    std::uint32_t proc_err = inf.getErrors();
+
+    // icu doesn't offer a flag to disable VerifyDnsLength or CheckHyphens, so just ignore those failures
+    proc_err &= ~UIDNA_ERROR_LEADING_HYPHEN;
+    proc_err &= ~UIDNA_ERROR_TRAILING_HYPHEN;
+    proc_err &= ~UIDNA_ERROR_HYPHEN_3_4;
+
+    if (!be_strict) {
+        proc_err &= ~UIDNA_ERROR_LABEL_TOO_LONG;
+        proc_err &= ~UIDNA_ERROR_DOMAIN_NAME_TOO_LONG;
+    }
+
+    // If domain or any label is empty, proc_err should contain UIDNA_ERROR_EMPTY_LABEL
+    if (U_FAILURE(err) || proc_err != 0) {
+        validation_error(ValidationError::DomainToASCII);
+
+        return std::nullopt;
+    }
+
+    return ascii_domain;
+}
+
+// https://url.spec.whatwg.org/#start-with-a-windows-drive-letter
+bool UrlParser::starts_with_windows_drive_letter(std::string_view input) {
+    if (input.size() >= 2) {
+        if (util::is_alpha(input[0]) && (input[1] == ':' || input[1] == '|')) {
+            if (input.size() == 2 || input[2] == '/' || input[2] == '\\' || input[2] == '?' || input[2] == '#') {
```

Comment:
You only know that input.size() is at least 2, so you can't access input[2] here.